File: ProblemSolve/1249.cs
/********************************************
 * Minimum Remove to Make Valid Parentheses *
 ********************************************/

public class Solution {
    public string MinRemoveToMakeValid(string s) {
        Stack<int> st = new Stack<int>();
        StringBuilder ans = new StringBuilder();
        int idx = 0;
        
        //모든 char을 돌면서 지워야 하는 괄호의 idx만 스택에 남긴다
        //ans에 하나씩 추가하면서 진행하기 때문에 ans를 기준으로 idx를 체크한다.
        foreach(char ch in s){
            if(ch == '('){
                ans.Append(ch);
                st.Push(idx++);
            }
            else if(ch == ')'){
                if(st.Any()){
                    ans.Append(ch);
                    st.Pop();
                    ++idx;
                }
            }
            else{
                ans.Append(ch);
                ++idx;
            }
        }        
        
        //여는 괄호가 남아있는 경우 지워준다.
        while(st.Any()){
            idx = st.Pop();
            ans.Remove(idx, 1);
        }
        
        return ans.ToString();
    }
    
    public string MinRemoveToMakeValid(string s) {
        StringBuilder ans = new StringBuilder();
        int bracketCount = 0;
        char[] input = s.ToCharArray();
        
        for(int i=0; i<input.Length; ++i){
            if(input[i] == '('){
                bracketCount++;
            }
            else if(input[i] == ')'){
                if(bracketCount == 0){
                    input[i] = 'D';
                }
                else{
                    bracketCount--;
                }
            }
        }
        
        bracketCount = 0;
        
        for(int i=input.Length-1; i>=0; --i){
            if(input[i] == ')'){
                bracketCount++;
            }
            else if(input[i] == '('){
                if(bracketCount == 0){
                    input[i] = 'D';
                }
                else{
                    bracketCount--;
                }
            }
        }
        
        foreach(char ch in input){
            if(ch != 'D'){
                ans.Append(ch);
            }
        }
        
        return ans.ToString();
    }
}

File: ProblemSolve/1249.cpp
/********************************************
 * Minimum Remove to Make Valid Parentheses *
 ********************************************/

class Solution {
public:
    string minRemoveToMakeValid(string s) {
        stack<int> st;
        int len = s.size();
        
        //지워야 하는 괄호를 'D'로 바꿔서 마지막에 지워준다
        for(int i=0; i<len; i++){
            if(s[i] == '('){
                st.push(i);
            }
            else if(s[i] == ')'){
                if(!st.empty()){
                    st.pop();
                }
                else{
                    s[i] = 'D';
                }
            }
        }
        
        while(!st.empty()){
            s[st.top()] = 'D';
            st.pop();
        }
        
        //'D'로 바뀐 char을 제거
        s.erase(remove(s.begin(), s.end(), 'D'), s.end());
        
        return s;
    }
    
    string minRemoveToMakeValid(string s) {
        int len = s.size();
        int bracketCount = 0;
        
        for(int i=0; i<len; ++i){
            if(s[i] == '('){
                bracketCount++;
            }
            else if(s[i] == ')'){
                if(bracketCount == 0){
                    s[i] = 'D';
                }
                else{
                    bracketCount--;
                }
            }
        }
        
        bracketCount = 0;
        
        for(int i=len-1; i>=0; --i){
            if(s[i] == ')'){
                bracketCount++;
            }
            else if(s[i] == '('){
                if(bracketCount == 0){
                    s[i] = 'D';
                }
                else{
                    bracketCount--;
                }
            }
        }
        
        string ans = "";
        
        for(char ch : s){
            if(ch != 'D')
                ans += ch;
        }
        
        return ans;
    }
};